### DISCUSSION TITLE
Welcome to gateloop-transformer Discussions!

### DISCUSSION THREAD

**Discussion by lucidrains** [lucidrains/gateloop-transformer#1] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by lucidrains** [comment]
thoughts about paper?

**Reply by lucidrains** [reply]
ok, i think i'm seeing a positive result.

edit: no longer seeing it

**Reply by jonmorton** [reply]
One interesting thing in the paper is the very high weight decay (0.25), haven't seen a value that high before. Not sure if it's important though.

**Reply by jonmorton** [reply]
Another notable difference it seems it that you are using pre-norm RMSNorm, whereas the paper is showing post-norm LN. Perhaps with gateloop there isn't as much instability with using post-norm.

**Reply by lucidrains** [reply]
yea that is very high, threw in weight decay support into the training script

i'll add post-ln support tomorrow

**Reply by lucidrains** [reply]
i'm actually surprised max-head works as well as it does

is it necessary to have keys and values projected just to elementwise multiply them? why couldn't that just be one projection? so many unanswered questions here

**Reply by tobiaskatsch** [reply]
As an additional hint, we used group norm and a swish gate following the RetNet paper. 
y = self.group_norm_function(y)
y = y * nn.swish(self.g_proj(x))
y = self.o_proj(y)
y = y.real

**Reply by lucidrains** [reply]
@tobiaskatsch 
<img width="745" alt="Screen Shot 2023-11-09 at 6 57 21 PM" src="https://github.com/lucidrains/gateloop-transformer/assets/108653/651d3c10-ff1c-488b-86c7-d79509720e8d">

yes, this is sub-ln design, which is from the same authors. i've tried sub-ln on regular transformer but didn't see anything remarkable, but in the 'retention' setting may be different

i'll try it out tomorrow morning, thanks!

**Reply by lucidrains** [reply]
how would groupnorm make sense for gateloop if each head dimension is reduced to 1?

**Reply by jonmorton** [reply]
Isn't the swish/group norm applied to the attention output after the heads are merged? RetNet shows: `GroupNorm(Concat(heads))`.

Edit: I guess you'd have to pick a GN group size that's != num heads

& Perhaps the use of group norm here is really what allows you to use post-norm ..

**Comment by jonmorton** [comment]
In the updated readme you mention:

> Also, gate loop's associative scan also is not able to train on even sequence lengths of 128

Is this at all because of how you had to implement it in pytorch? I.e., would you expect better results in jax with its native associate scan? The paper is also suspiciously missing any mention of training compute comparisons, so maybe it's a fundamental issue.

**Reply by lucidrains** [reply]
in jax there would be the same problem

this has been a long standing issue with all autoregressive formulations of linear attention and requires [a CUDA kernel](https://github.com/idiap/fast-transformers/blob/master/fast_transformers/causal_product/causal_product_cuda.cu) to avoid materializing the [`kv` intermediate](https://github.com/lucidrains/gateloop-transformer/blob/main/gateloop_transformer/gateloop_transformer.py#L136) of dimensions `(batch, seq, dim_qk, dim_v)`

**Reply by tobiaskatsch** [reply]
Hello, I am the author. For the experiments, we used JAX parallel scan. In theory, the shape is (B, L, H, D, D), where H is the number of heads and D is the hidden state size. In practice, this will pose a memory bottleneck. We avoid this by choosing a max-head variant, e.g., (D=1), meaning that kv is of shape (B, L, H, 1, 1), coinciding with element-wise gating. The more general case was mainly included in the paper to highlight the connection to standard multi-head attention. This should have been explained more clearly and will be addressed in the upcoming version.

**Reply by lucidrains** [reply]
I see! Thank you for clarifying

I will rerun the experiments 🙏

**Reply by tobiaskatsch** [reply]
Also, from my experience, the quadratic O(l**2) formulation is not practical (for now) because using the cumulative product of magnitudes is numerically unstable and obvious fixes (like the clamping you try) is not sufficient in pracise. All our experience were conducted using the parallel scan variant.

**Reply by lucidrains** [reply]
Thank you

**Comment by jrb314** [comment]
Awesome work! 
Any plan to replicate the main wikitext103 result in the paper (13.4 perplexity)?
Thanks!

**Reply by lucidrains** [reply]
i'm running experiments right now on enwik8, a dataset i'm familiar with. the experiments are all public (link in readme) and will grow throughout today and tomorrow.

feel free to conduct your own experiments and share them. i have everything setup in `train.py`

**Reply by lucidrains** [reply]
i'm very curious how the author was able to conduct wikitext103 on sufficient sequence lengths without the proper cuda kernel. the only way for this scheme to work is to fuse everything together, as we know from [autoregressive linear attention line of work](https://arxiv.org/abs/2006.16236)

**Reply by lucidrains** [reply]
but yea, let's see how it holds up to small scale experiments

**Reply by jrb314** [reply]
Paging @tobiaskatsch.  If you are the paper's author, maybe you can shed some light?

**Reply by tobiaskatsch** [reply]
See https://github.com/lucidrains/gateloop-transformer/discussions/1#discussioncomment-7516938

**Comment by AlexDeMoor** [comment]
The official code has been released : https://github.com/tobiaskatsch/GateLoop

**Reply by lucidrains** [reply]
@AlexDeMoor nice, somebody check out the wikitext103 experiments, make sure there is no test leakage

**Reply by lucidrains** [reply]
@AlexDeMoor what happened with this? anybody ran it?
